adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m

```diff
@@ -0,0 +1,220 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GADPangleRTBNativeRenderer.h"
+#import "GADMediationAdapterPangleConstants.h"
+#import <BUAdSDK/BUAdSDK.h>
+#import "GADMAdapterPangleUtils.h"
+#include <stdatomic.h>
+
+static NSString *const BUDNativeAdTranslateKey = @"bu_nativeAd";
+
+@interface GADPangleRTBNativeRenderer()<BUNativeAdDelegate> {
+    /// The completion handler to call when the ad loading succeeds or fails.
+    GADMediationNativeLoadCompletionHandler _loadCompletionHandler;
+    /// The Pangle native ad.
+    BUNativeAd *_nativeAd;
+    /// The Pangle related view.
+    BUNativeAdRelatedView *_relatedView;
+    /// An ad event delegate to invoke when ad rendering events occur.
+    id<GADMediationNativeAdEventDelegate> _delegate;
+}
+
+@end
+
+@implementation GADPangleRTBNativeRenderer
+@synthesize images = _images,icon = _icon, mediaView = _mediaView;
+
+- (void)renderNativeAdForAdConfiguration:(nonnull GADMediationNativeAdConfiguration *)adConfiguration
+                       completionHandler:(nonnull GADMediationNativeLoadCompletionHandler)completionHandler {
+    __block atomic_flag completionHandlerCalled = ATOMIC_FLAG_INIT;
+    __block  GADMediationNativeLoadCompletionHandler originalCompletionHandler = [completionHandler copy];
+    _loadCompletionHandler = ^id<GADMediationNativeAdEventDelegate> (_Nullable id<GADMediationNativeAd> ad, NSError *_Nullable error) {
+        if (atomic_flag_test_and_set(&completionHandlerCalled)) {
+            return nil;
+        }
+        id<GADMediationNativeAdEventDelegate> delegate = nil;
+        if (originalCompletionHandler) {
+            delegate = originalCompletionHandler(ad, error);
+        }
+        originalCompletionHandler = nil;
+        return delegate;
+    };
+    
+    NSString *placementId = adConfiguration.credentials.settings[GADMAdapterPanglePlacementID] ?: @"";
+    if (!placementId.length) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters,
+                                                                      [NSString stringWithFormat:@"%@ cannot be nil.",GADMAdapterPanglePlacementID]
+                                                                      );
+        _loadCompletionHandler(nil, error);
+        return;
+    }
+    
+    BUAdSlot *slot = [[BUAdSlot alloc] init];
+    slot.ID = placementId;
+    slot.AdType = BUAdSlotAdTypeFeed;
+    
+    _nativeAd = [[BUNativeAd alloc]initWithSlot:slot];
+    if (!_nativeAd) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInitAd,
+                                                                      @"Init ad error, Please check the reason or you could contact Pangle"
+                                                                      );
```

Comment:
Why is this needed for native but not for other ad formats?

For example for rewarded, we have the below, but we didn't do the same check:
```
  _rewardedVideoAd = [[BURewardedVideoAd alloc] initWithSlotID:placementId
                                            rewardedVideoModel:model];
```

Reply:
@wangyuanyuanyy to respond. Thanks!

Reply:
If Pangle SDK does this check internally (and fails under this condition), we can remove this check.